### ISSUE TITLE
Window function being applied after LIMIT.

### ISSUE THREAD

**Issue by wol-e** [ClickHouse/ClickHouse#23125]
Hi. I noticed that when using window functions with a LIMIT clause, the window function appears to be applied to the result _after_ applying the limit. To me this is rather unintuitive and different from the way I am used to it from other applications. 
Is this intended behaviour or should I raise an issue for this? 

Example (clickhouse version 21.3.5.42):

````
CREATE TABLE test.window_test ENGINE = MergeTree()
PARTITION BY field
ORDER BY field
SETTINGS index_granularity = 8192
AS
SELECT
    'a' AS field
UNION ALL
SELECT
    'a' AS field
UNION ALL
SELECT
    'b' AS field
UNION ALL
SELECT
    'b' AS field
    ;

set allow_experimental_window_functions = 1;

-- Example query with expected output:
SELECT
    field,
    count() over (partition by field) AS window_count
FROM
    test.window_test;

-- Output 
--| field | window_count | 
--| a      | 2      | 
--| a      | 2      | 
--| b      | 2      | 
--| b      | 2      |

-- Query with LIMIT giving unexpected output:
SELECT
    field,
    count() over (partition by field) AS window_count
FROM
    test.window_test
LIMIT 1;

-- Output
--| field | window_count | 
--| a      | 1       | 

-- Expected Output:
--| field | window_count | 
--| a      | 2      | 
````

**Comment by akuzm** [comment]
This is a bug, `limit` shouldn't be moved under the calculation of window functions. It's probably not very important, because the result of `limit` w/o `order by` is unspecified, and if you add an explicit `order by`, it works as expected:
```
WITH arrayJoin(['a', 'a', 'b', 'b']) AS field
SELECT
    field,
    count() OVER (PARTITION BY field)
ORDER BY field ASC
LIMIT 1

┌─field─┬─count() OVER (PARTITION BY arrayJoin(['a', 'a', 'b', 'b']) AS field)─┐
│ a     │                                                                    2 │
└───────┴──────────────────────────────────────────────────────────────────────┘
```

**Comment by wol-e** [comment]
Thanks for the quick response and the simple work around 👍

**Comment by alexey-milovidov** [comment]
@akuzm Would you like to address this issue or we'd better close as "won't fix"?

**Comment by akuzm** [comment]
> @akuzm Would you like to address this issue or we'd better close as "won't fix"?

The use case is of minor importance and has a simple workaround, but it is still a correctness problem, so I wouldn't close with "won't fix". I think it will be fixed as a side effect when we redo how we push down sorting nodes in the execution plan (@KochetovNicolai has some plans for this).

**Comment by den-crane** [comment]
I've found the same issue

```
select * from (　( select *, count() over () cnt from　( select * from numbers(10000000) )　)　)  limit 3　;
┌─number─┬─cnt─┐
│      0 │   3 │
│      1 │   3 │
│      2 │   3 │
└────────┴─────┘

select * from (　( select *, count() over () cnt from　( select * from numbers(10000000) )　)　) order by number limit 3　;
┌─number─┬──────cnt─┐
│      0 │ 10000000 │
│      1 │ 10000000 │
│      2 │ 10000000 │
└────────┴──────────┘
```

**Comment by filimonov** [comment]
```
SELECT
    number,
    leadInFrame(number) OVER w AS W
FROM numbers(10)
WINDOW w AS (ORDER BY number ASC Rows BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
LIMIT 3;

┌─number─┬─W─┐
│      0 │ 1 │
│      1 │ 2 │
│      2 │ 0 │ <-- W should be 3!
└────────┴───┘
```

In contrast:  https://www.db-fiddle.com/f/hNNECnp7YC3E2zBKBaxN6T/1

**Comment by novikd** [comment]
Fixed in #36075
